stdlib/public/core/StringWordView.swift

```diff
@@ -0,0 +1,120 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension String {
+  internal struct _WordView {
+    internal var _guts: _StringGuts
+
+    internal init(_ _guts: _StringGuts) {
+      self._guts = _guts
+    }
+  }
+}
+
+extension String._WordView: Collection {
+  internal struct Index {
+    @usableFromInline
+    internal var _encodedOffset: Int
+  }
+
+  typealias Element = Substring
+
+  var startIndex: Index {
+    Index(_encodedOffset: _guts.startIndex._encodedOffset)
+  }
+
+  var endIndex: Index {
+    Index(_encodedOffset: _guts.endIndex._encodedOffset)
+  }
+
+  func index(after i: Index) -> Index {
+    _internalInvariant(i < endIndex)
+
+    let nextOffset = _guts._opaqueNextWordIndex(startingAt: i._encodedOffset)
+    return Index(_encodedOffset: nextOffset)
+  }
+
+  subscript(position: Index) -> Element {
+    let indexAfter = index(after: position)
+
+    let l = String.Index(_encodedOffset: position._encodedOffset)
+    let u = String.Index(_encodedOffset: indexAfter._encodedOffset)
+
+    return String(_guts)[l ..< u]
+  }
+}
+
+extension String._WordView.Index: Equatable {
+  @inlinable
+  @inline(__always)
+  static func == (lhs: Self, rhs: Self) -> Bool {
+    lhs._encodedOffset == rhs._encodedOffset
+  }
+}
+
+extension String._WordView.Index: Comparable {
+  @inlinable
+  @inline(__always)
+  static func < (lhs: Self, rhs: Self) -> Bool {
+    lhs._encodedOffset < rhs._encodedOffset
+  }
+}
+
+extension String._WordView.Index {
+  internal init(_ i: String.Index) {
+    _encodedOffset = i._encodedOffset
+  }
+}
+
+extension String._WordView.Index: Sendable {}
+
+extension String._WordView: BidirectionalCollection {
+  func index(before i: Index) -> Index {
+    _internalInvariant(i > startIndex)
```

Comment:
Why no up front check against the endIndex? Why is this only an `_internalInvariant`?

I'd very much like to avoid another case of https://github.com/apple/swift/pull/41598 here.